### ISSUE TITLE
libcsp-1 backwards compatibility

### ISSUE THREAD

**Issue by n-log-gos** [libcsp/libcsp#724]
## Description

How can I achieve compatibility between a _node A_ using **libcsp-2.0 + csp_conf.version=1**  and a _node B_ that uses **libcsp-1** ? I can't access the source code of _node B_ so upgrading isn't possible. 

`csp_packet_t` structure is inherently different between each version, the most prominent difference being the fixed big-endianness of the header of the newer versions. Is there an effort already done to create this compatibility?

## Steps to Reproduce

Compile a CSP server and client from libcsp-2.0 and libcsp-1, communication will fail.

## Environment Details

Provide information about your environment, including:

- Build OS:  Ubuntu 24.04
- Target OS: Ubunut 24.04
- libcsp commit hash: f74ee417 vs. 7fba4018

**Comment by yashi** [comment]
Does #713 help?

**Comment by n-log-gos** [comment]
> Does #713 help?

I have been using `csp_conf.version=1` - via the `-v` option referenced in the PR. However, this does not entirely work, as what I need is to use a "libcsp-2.0 node" (with `-v1` option to set CSP version 1) and a "libcsp-1" node, i.e., a propietary software which source code / upgrade I can't access to. 

My questions is: is there a test mode where I could set libcsp-2 to be compatible with libcsp-1 nodes? I.e., nodes that are compatible with the old, libcsp-1 `csp_packet_t` struct ? Or should I downgrade and just use libcsp-1 for all my nodes ?

**Comment by yashi** [comment]
We also have a Cubesat component runing CSPv1. And we have successfully communicating using `develop` branch of libcsp in CSP protocol v1.

**Comment by yashi** [comment]
Are you able to capture the packets using Wireshark? Does https://github.com/KantaTamura/CSP-dissector help?

**Comment by n-log-gos** [comment]
> We also have a Cubesat component runing CSPv1. And we have successfully communicating using `develop` branch of libcsp in CSP protocol v1.

Is that Cubesat component built using libcsp-1 library? Or is built against libcsp2.0, and using `csp_conf = 1 `?

**Comment by yashi** [comment]
I’m not sure; it’s not disclosed. However, we bought it long before we tagged v2, so it must be running libcsp-v1.

I just checked out the `libcsp-1` branch and tested it against the `develop` branch with protocol v1 via USART on PTY—it seems to work fine.  Tell me how exactly you have done to make them fail.

**Comment by n-log-gos** [comment]
Sure, let me tell you how exactly I do this. I use ZMQ proxy, in my case.

First, launch a **libcsp1** based CSP client:
```shell
./csp_server_client -a11 -r10 -d6 -zlocalhost
``` 
Then, launch a **libcsp2** based CSP server and ZMQ proxy, in both cases using `-v1:`
```shell
./csp_server -zlocalhost -a10 -v1
./zmqproxy -v1 -d
```

You would expect the ZMQ proxy logger to show CSP addr src = 11 and dst = 10 and the CSP server to receive the ping + Hello World messages. However, I get no pings at the server side. Moreover, the ZMQ proxy logs:

```shell
Packet: Src 5, Dst 0, Dport 1, Sport 40, Pri 0, Flags 0xA0, Size 101
Packet: Src 10, Dst 5, Dport 40, Sport 1, Pri 0, Flags 0xA0, Size 101
``` 

Which indicates that at some point the message was not decoded correctly.

**Comment by yashi** [comment]
Ah, ZMQ....

Is your _Node B_ using ZMQ? On Linux? Why can't you update it?

**Comment by n-log-gos** [comment]
> Is your _Node B_ using ZMQ? On Linux? Why can't you update it?

Yes, it uses ZMQ proxy interface on Linux Ubuntu 24.04. I can't update it because I don't have access to the source code of that software (softmodem). 

I believe that CSP over KISS works OK when mixing _libcsp1_ and _libcsp2_ nodes because the old libcsp1 `csp_kiss_rx / csp_kiss_tx` functions converted the CSP header to big endian. But that's not happening in `csp_zmq_rx / csp_zmq_tx`, where the header was sent "as is".

**Comment by yashi** [comment]
Would you mind creating a PR for it?

**Comment by n-log-gos** [comment]
sure, working on that

**Comment by nml0** [comment]
I'm not very happy with all the changes needed for this.
My proposal is: https://github.com/libcsp/libcsp/pull/738

Please take a look at this. For us, this patch is working fine and let us have backward compatibility.

**Comment by yashi** [comment]
@nml0 I really appreciate your work! Thanks a lot!

**Comment by yashi** [comment]
#738 has been merged. closing.
